parser_library/test/context/macro_processing_stack_test.cpp

```diff
@@ -0,0 +1,454 @@
+{
+    mock_parse_lib_provider lib({
+        { "ACTION", " MNOTE 'Hello'" },
+        { "COPYBOOK", R"(
+    MACRO
+    MAC
+    COPY ACTION
+    MEND
+)" },
+    });
+    std::string input = R"(
+    COPY COPYBOOK
+    MAC
+)";
+    analyzer a(input, analyzer_options { opencode, &lib });
+    a.analyze();
+
+    ASSERT_TRUE(matches_message_codes(a.diags(), { "MNOTE" }));
+    EXPECT_TRUE(matches_diagnostic_stack(a.diags().front(), { "ACTION", "COPYBOOK", "opencode" }));
+}
+
+TEST(macro_processing_stack, copy_whole_macro_with_nested_macro_in_copybook)
+{
+    mock_parse_lib_provider lib({
+        { "NESTCPY", R"(
+    MACRO
+    NESTED
+    MNOTE 'Hello'
+    MEND
+)" },
+        { "COPYBOOK", R"(
+    MACRO
+    MAC
+    COPY NESTCPY
+    MEND
+)" },
+    });
+    std::string input = R"(
+    COPY COPYBOOK
+    MAC
+    NESTED
+)";
+    analyzer a(input, analyzer_options { opencode, &lib });
+    a.analyze();
+
+    ASSERT_TRUE(matches_message_codes(a.diags(), { "MNOTE" }));
+    EXPECT_TRUE(matches_diagnostic_stack(a.diags().front(), { "NESTCPY", "opencode" }));
+}
+
+TEST(macro_processing_stack, copy_whole_macro_with_nested_macro_in_copybook_with_copybook)
+{
+    mock_parse_lib_provider lib({
+        { "ACTION", " MNOTE 'Hello'" },
+        { "NESTCPY", R"(
+    MACRO
+    NESTED
+    COPY ACTION
+    MEND
+)" },
+        { "COPYBOOK", R"(
+    MACRO
+    MAC
+    COPY NESTCPY
+    MEND
+)" },
+    });
+    std::string input = R"(
+    COPY COPYBOOK
+    MAC
+    NESTED
+)";
+    analyzer a(input, analyzer_options { opencode, &lib });
+    a.analyze();
+
+    ASSERT_TRUE(matches_message_codes(a.diags(), { "MNOTE" }));
+    EXPECT_TRUE(matches_diagnostic_stack(a.diags().front(), { "ACTION", "NESTCPY", "opencode" }));
+}
+
+TEST(macro_processing_stack, double_copy_with_nested_0)
+{
+    mock_parse_lib_provider lib({
+        { "COPY1", " COPY COPY2" },
+        { "COPY2", R"(
+    MACRO
+    MAC
+    MACRO
+    NESTED
+    MACRO
+    INNER
+    MNOTE 'Hello'
+    MEND
+    MEND
+    MEND
+)" },
+    });
+    std::string input = R"(
+    COPY COPY1
+    MAC
+    NESTED
+    INNER
+)";
+    analyzer a(input, analyzer_options { opencode, &lib });
+    a.analyze();
+
+    ASSERT_TRUE(matches_message_codes(a.diags(), { "MNOTE" }));
+    EXPECT_TRUE(matches_diagnostic_stack(a.diags().front(), { "COPY2", "opencode" }));
+}
+
+TEST(macro_processing_stack, double_copy_with_nested_1)
+{
+    mock_parse_lib_provider lib({
+        { "COPY1", " COPY COPY2" },
+        { "COPY2", R"(
+    MACRO
+    NESTED
+    MACRO
+    INNER
+    MNOTE 'Hello'
+    MEND
+    MEND
+)" },
+        { "MAC", R"(.*
+    MACRO
+    MAC
+    COPY COPY1
+    MEND
+)" },
+    });
+    std::string input = R"(
+    MAC
+    NESTED
+    INNER
+)";
+    analyzer a(input, analyzer_options { opencode, &lib });
+    a.analyze();
+
+    ASSERT_TRUE(matches_message_codes(a.diags(), { "MNOTE" }));
+    EXPECT_TRUE(matches_diagnostic_stack(a.diags().front(), { "COPY2", "opencode" }));
```

Comment:
Maybe we can discuss this over lunch but I'm not sure I get why `MAC` is not part of the list. I'm probably missing something :)

Reply:
My thinking is that the whole definition is contained in the `COPY2` file.
So when the `INNER` macro is entered, it starts in file `COPY2`.

Reply:
What I think I'm struggling with is that I find the `nested_macro_with_copy` test similar, yet the diag stack works a bit differently. But let's discuss :)

Reply:
The difference there is that the NESTED macro actually starts in the `MAC` file, but then the MNOTE instruction is in the `COPYBOOK` file.

Reply:
I see, makes sense. Thanks!